flask/src/main/java/net/corda/gradle/flask/FlaskJarTask.java

```diff
@@ -0,0 +1,235 @@
+        mainClassName = objects.property(String.class);
+        jvmArgs = objects.listProperty(String.class).convention(new ArrayList<>());
+        javaAgents = new ArrayList<>();
+        from(getProject().tarTree(LauncherResource.instance), copySpec -> exclude(JarFile.MANIFEST_NAME));
+    }
+
+    @Input
+    @SneakyThrows
+    public String getLauncherArchiveHash() {
+        MessageDigest md5 = MessageDigest.getInstance("MD5");
+        try(InputStream stream = new DigestInputStream(LauncherResource.instance.read(), md5)) {
+            byte[] buffer = new byte[0x10000];
+            while(stream.read(buffer) >= 0) {}
+        }
+        return String.format("%032x", new BigInteger(1, md5.digest()));
+    }
+
+    @RequiredArgsConstructor
+    private static class StreamAction implements CopyActionProcessingStreamAction {
+
+        private final ZipOutputStream zoos;
+        private final Manifest manifest;
+        private final MessageDigest md;
+        private final byte[] buffer;
+
+        private final List<FileCopyDetailsInternal> fileCopyDetailsInternals = new ArrayList<>();
+
+        @SneakyThrows
+        public void write() {
+            ZipEntry zipEntry = new ZipEntry(Flask.Constants.METADATA_FOLDER + '/');
+            zoos.putNextEntry(zipEntry);
+            zipEntry = new ZipEntry(JarFile.MANIFEST_NAME);
+            zipEntry.setMethod(ZipEntry.DEFLATED);
+            zoos.putNextEntry(zipEntry);
+            manifest.write(zoos);
+            zoos.closeEntry();
+            for(FileCopyDetails fileCopyDetails : fileCopyDetailsInternals) {
+                String entryName = fileCopyDetails.getRelativePath().toString();
+                if(Objects.equals(Flask.Constants.METADATA_FOLDER, entryName)) continue;
+                zipEntry = new ZipEntry(entryName);
+                if(fileCopyDetails.isDirectory()) {
+                    zipEntry = new ZipEntry(entryName + '/');
+                    zoos.putNextEntry(zipEntry);
+                } else {
+                    boolean compressed = Flask.splitExtension(fileCopyDetails.getSourceName())
+                            .map(entry -> Objects.equals(".jar", entry.getValue()))
+                            .orElse(false);
+                    if (!compressed) {
+                        zipEntry.setMethod(ZipEntry.DEFLATED);
+                    } else {
+                        try (InputStream is = new FileInputStream(fileCopyDetails.getFile())) {
+                            Flask.computeSizeAndCrc32(zipEntry, is, buffer);
+                        }
+                        zipEntry.setMethod(ZipEntry.STORED);
+                    }
+                    try(InputStream is = new FileInputStream(fileCopyDetails.getFile())) {
+                        zoos.putNextEntry(zipEntry);
+                        Flask.write2Stream(zoos, is, buffer);
+                        zoos.closeEntry();
```

Comment:
The `zoos.closeEntry()` is implicit in the `zoos.putNextEntry()` and `zoos.close()` calls.

Reply:
ok, will remove